testrunner

```diff
@@ -266,35 +266,87 @@ install_packages() {
     fi
 }
 
-run_tests() {
-    local upload_dir=$(cat $state/upload_dir)
-
-    update_github pending "Running tests."
-
-    local rc=0
-
-    max_date=$(($(date +%s) + $(cat $state/parsed/timeout)))
+run_loop() {
+    local timeout=$1; shift
+    local logfile=$1; shift
+    local workdir=$1; shift
+    local testfile=$1; shift
+    local envfile=$1; shift
 
+    local max_date=$(($(date +%s) + $timeout))
     while IFS='' read -r line || [[ -n $line ]]; do
 
-        local timeout=$(($max_date - $(date +%s)))
-
+        timeout=$(($max_date - $(date +%s)))
         if [ $timeout -le 0 ]; then
-            echo "### TIMED OUT" >> $upload_dir/output.log
+            echo "### TIMED OUT" >> $logfile
             rc=137
             break
         fi
 
         rc=0
-        logged_envcmd \
-            $upload_dir/output.log /var/tmp/checkout \
-            $state/parsed/envs $timeout "$line" || rc=$?
+        logged_envcmd $logfile $workdir $envfile $timeout "$line" || rc=$?
 
         if [ $rc != 0 ]; then
             break
         fi
 
-    done < $state/parsed/tests
+    done < "$testfile"
+
+    return $rc
+}
+
+build_and_test() {
+    local upload_dir=$(cat $state/upload_dir)
+    local timeout=$(cat $state/parsed/timeout)
+    local checkout=checkouts/$github_repo
+    local rc=0
+
+    if [ -f $state/parsed/build ]; then
+        local config_opts=$(cat $state/parsed/build.config_opts)
+        local build_opts=$(cat $state/parsed/build.build_opts)
+        local install_opts=$(cat $state/parsed/build.install_opts)
+
+        update_github pending "Building..."
+
+        touch $state/build.sh
+        if [ ! -f $checkout/configure ]; then
+            if [ -f $checkout/autogen.sh ]; then
+                echo "NOCONFIGURE=1 ./autogen.sh" >> $state/build.sh
+            elif [ -f $checkout/autogen ]; then
+                echo "NOCONFIGURE=1 ./autogen" >> $state/build.sh
+            fi
+        fi
+
+        echo "./configure $config_opts" >> $state/build.sh
```

Comment:
I'd consider doing separate `srcdir != builddir` by default, like GContinuous does.  It makes it easier in the future to parallelize builds of the same source.  (debug/optimized etc).